packages/@aws-cdk/aws-sns/lib/subscription.ts

```diff
@@ -215,3 +253,31 @@ export enum SubscriptionProtocol {
    */
   FIREHOSE = 'firehose'
 }
+
+/**
+ * The filter policy after it has had filters transformed by SubscriptionFilter
+ */
+interface SubscriptionFilterPolicyArrayV2 {
+  [attribute: string]: any[] | SubscriptionFilterPolicyArrayV2;
+}
+
+/**
+ * The filter policy that allows for nested subscription filter policies. This is accomplished via recursive types.
+ */
+export interface SubscriptionFilterPolicyV2 {
+  [attribute: string]: SubscriptionFilter | SubscriptionFilterPolicyV2;
+}
```

Comment:
I think we should try to avoid using a union type to model this, but I can't think of a better way. https://aws.github.io/jsii/specification/2-type-system/#type-unions